src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs

```diff
@@ -0,0 +1,142 @@
+            if (syntaxFacts.IsInNonUserCode(context.SyntaxTree, context.Position, cancellationToken) ||
+                syntaxFacts.IsSkippedTokensTrivia(node)) // Note: IsSkippedTokensTrivia implementation is probably buggy
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            if (context.TargetToken.RawKind == syntaxFacts.SyntaxKinds.DotToken)
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var span = new TextSpan(position, 0);
+            var enumType = typeInferenceService.InferType(context.SemanticModel, position, objectAsDefault: true, cancellationToken: cancellationToken);
+
+            if (enumType is not { TypeKind: TypeKind.Enum })
+            {
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+            }
+
+            var hideAdvancedMembers = options.GetOption(RecommendationOptions.HideAdvancedMembers, context.SemanticModel.Language);
+
+            var otherSymbols = context.SemanticModel.LookupSymbols(position).WhereAsArray(s =>
+                s.MatchesKind(SymbolKind.Field, SymbolKind.Local, SymbolKind.Parameter, SymbolKind.Property) &&
+                s.IsEditorBrowsable(hideAdvancedMembers, context.SemanticModel.Compilation));
+
+            var otherInstances = otherSymbols.WhereAsArray(s => Equals(enumType, s.GetSymbolType()));
+
+            return Task.FromResult(otherInstances.Concat(enumType));
+        }
+
+        // PERF: Cached values for GetDisplayAndInsertionText. Cuts down on the number of calls to ToMinimalDisplayString for large enums.
+        private (INamedTypeSymbol? containingType, SyntaxContext? context, string? containingTypeText) _cachedDisplayAndInsertionText;
+
+        protected override (string displayText, string suffix, string insertionText) GetDisplayAndSuffixAndInsertionText(ISymbol symbol, SyntaxContext context)
+        {
+            if (symbol.ContainingType != null && symbol.ContainingType.TypeKind == TypeKind.Enum)
+            {
+                if (!Equals(_cachedDisplayAndInsertionText.containingType, symbol.ContainingType) || _cachedDisplayAndInsertionText.context != context)
+                {
+                    var displayFormat = SymbolDisplayFormat.MinimallyQualifiedFormat
+                        .WithMemberOptions(SymbolDisplayMemberOptions.IncludeContainingType)
+                        .WithLocalOptions(SymbolDisplayLocalOptions.None);
+                    var containingTypeText = symbol.ContainingType.ToMinimalDisplayString(context.SemanticModel, context.Position, displayFormat);
+                    _cachedDisplayAndInsertionText = (symbol.ContainingType, context, containingTypeText);
```

Comment:
this doesn't seem threadsafe at all.

Reply:
Yes. I could wrap it with `ReaderWriterLockSlim` or remove the optimization altogether. What should I do?

Reply:
let's wrap for now.

Reply:
I did but used a `lock` because I needed to use `ReaderWriterLockSlim.EnterUpgradeableReadLock` which basically behaves like `lock`.